lib/elixir/lib/inspect/algebra.ex

```diff
@@ -354,28 +347,24 @@ defmodule Inspect.Algebra do
             try do
               Process.put(:inspect_trap, true)
 
-              res =
-                Inspect.Map.inspect(struct, %{
-                  opts
-                  | syntax_colors: [],
-                    inspect_fun: Inspect.Opts.default_inspect_fun()
-                })
-
-              res = IO.iodata_to_binary(format(res, :infinity))
-
-              message =
-                "got #{inspect(caught_exception.__struct__)} with message " <>
-                  "#{inspect(Exception.message(caught_exception))} while inspecting #{res}"
-
-              exception = Inspect.Error.exception(message: message)
+              error_exception =
+                Inspect.Error.exception(
+                  exception: caught_exception,
+                  inspect_opts: opts,
+                  stacktrace: __STACKTRACE__,
+                  struct: struct
+                )
 
               if opts.safe do
-                Inspect.inspect(exception, %{
-                  opts
-                  | inspect_fun: Inspect.Opts.default_inspect_fun()
-                })
+                opts = %{opts | inspect_fun: Inspect.Opts.default_inspect_fun()}
+                Inspect.inspect(error_exception, opts)
               else
-                reraise(exception, __STACKTRACE__)
+                attributes =
+                  error_exception
+                  |> Map.drop([:__struct__, :__exception__])
+                  |> Map.to_list()
+
+                reraise(Inspect.Error, attributes, __STACKTRACE__)
```

Comment:
We should not change this behaviour. `safe` should always return the original exception.

Reply:
I am confused about this comment, because this section is for safe=false,
anyway the previous code we return and raise Inspect.Error.
`exception = Inspect.Error.exception(message: message)`

Reply:
Sorry, my comment was confusing. I am saying you can do this: `reraise(error_exception, __STACKTRACE__)`, you don't need to extract and rebuild it with `reraise/3`.